Cookie Clicker in SwiftUI

Abizern opened the pull request:
Closes #16 

This is a very basic tutorial that essentially recreates the storyboard and UIKit version that already exists.

It's not very long, but from experience at previous events this is about as much as can be covered with a beginner in the time available, and there is scope for further tweaking and additions with the coach for those that finish faster.

Sorry it took so long - writing tutorials is hard. I want to put so much more into it, but I know that would be overwhelming.

**Any and all** input is welcome. I tried to match the tone of the previous tutorial, maybe too much, or maybe too little?

As this is a new chunk, I haven't done any of the work to run the site generator over the new tutorial. If this gets approved I can take a stab at that in a separate PR.

daniellevass commented on src/docs/cookie-clicker-swiftui.md:
````diff
@@ -0,0 +1,314 @@
+
+```swift
+struct ContentView: View {
+    @State private var showGreeting = false
````
is there any way we can do this without using @state variables? e.g. create a standard Int variable that holds how many times you've tapped the button. And then in the next section hook up the button to a regular method that increments (and displays?) that counter?

Abizern replied:
Maybe it's possible, but I wouldn't want to go that way. SwiftUI depends very much on the `@State`. It's what makes it a declarative way of writing views. It's not just `@State` (which is internal to a view) but also `@ObservedObject` `@Environment` and `@EnvironmentObject` which make this all much more powerful and I was hoping to cover those in subsequent tutorials - (I have a plan 😄 ).

Another reason I'd rather not change this is because If we are going to coach attendees on SwiftUI, I think it's better to teach them real world code, however simple, that's the benefit of having coaches, rather than throwing a tutorial over the wall. There are many tutorials about this, but I like that Codebar does this with a coach, who can explain things further from their own experience.

Does my reticence make sense in this context?

daniellevass replied:
Yeah! I think that definitely makes sense. 

Would you be able to explain what a @state variable is - or link to a place that explains well what it is? I love that we have 2:1 learning too, but some people use our tutorials at home on their own, or in different group settings too.

daniellevass commented on src/docs/cookie-clicker-swiftui.md:
````diff
@@ -0,0 +1,314 @@
+Change the state variable to hold a number instead of a `Bool`
+
+```swift
+@State private var count = 0
````
might be a stupid question, but what does @state add to a variable?

Abizern replied:
Ah. That's a really good point. @State is important. It's a property modifier that the `body` of the view uses to know that when it changes the view should redraw. If you didn't understand that (I know you'll take that phrase in the way I mean it), then I haven't explained it properly. I'll see if I can do an edit to make it clearer.

daniellevass replied:
Thank you 🎉

daniellevass commented on src/docs/cookie-clicker-swiftui.md:
```diff
@@ -0,0 +1,314 @@
+
+![step4](assets/cookie_clicker_swiftui/step4.gif)
+  
+If you click on the `Resume` button in the top right, and you are running Mac OS X Catalina you will see a preview of the **Hello World** application that the template has provided for you.
```
this took a few minutes for my macbook to load the first time with 8GB ram, is this something that might take some time for students to see too?

Abizern replied:
Yes. Usually it's because you've just installed or upgraded Xcode, or because you've chosen a simulator that you've never used before. Their is some initialisation to the simulator the first time. I know you're running in preview, but it sort of uses a simulator. Subsequent loads are much faster.

daniellevass replied:
great - i'm happy to know it's not just me 😆 ! We can let students know to expect this may take some time :)

Abizern commented:
> Thank you @Abizern for starting this - it's really a great start and you've taught me some things too!
> 
> I hope you don't mind that I changed some wording of the begining content.
> 
> I've also left a few comments in the code - is there a way we can do the greeting code using simpler code? I'm not sure many students will have experienced ternaries before, is there a way to use an if else instead?
> 
> Let me know if there's anything I can do to help 🎉

It is possible to use `if/else` as long as it's exhaustive. It's not possible to use `if let... / else` (there is much gnashing in the iOS-Sphere about this.

But yes, it may be out of a new developer's comfort zone. I'll see if I can make it simpler, and add rewriting as a ternary as one of the extra credit suggestions.

Abizern commented:
Thanks @daniellevass for taking the time to go through this. Perfectly happy for you to make edits directly.

You've made some really good points about clarity, I'll make some changes to this PR to address them.

Abizern commented:
@daniellevass I've Updated the tutorial to remove the ternary and added a bit more explanation around State variables. Please see if it makes it any clearer.

daniellevass commented:
I'm going to merge this in tonight so we can see it working with real students!
